nixos/modules/misc/version.nix

```diff
@@ -263,6 +263,19 @@ in
 
   config = {
 
+    assertions = [
+      {
+        assertion = builtins.match "[0-9]{2}\\.[0-9]{2}" config.system.stateVersion != null;
+        message = ''
+          ${config.system.stateVersion} is an invalid value for `system.stateVersion`; it must be in the format `YY.MM`, which corresponds to a prior release of NixOS.
+
+          If you want to switch releases or switch to unstable, you should change your channel and/or flake input URLs only.
+          *DO NOT* touch the `system.stateVersion` option, as it will not help you upgrade.
```

Comment:
Nit but since backticks aren't used for anything in terminals, I would like to suggest using `' '` instead. Looks a little more coherent imho.